### ISSUE TITLE
Align Assignments shortcut messes with square brackets on some keyboard layouts

### ISSUE THREAD

**Issue by jbaute** [Microsoft/VS-PPT#69]
Hi,

This is an old one, but I just ran into it again by installing the PPT plugin in VS2015.
What happens is that the shortcut for Align Assignments is set to Ctrl-Alt-^.
On a Belgian keyboard however, this is what you have to press to get an opening square bracket.
Let me tell you. It's rather annoying if that doesn't end up working anymore when you're typing code. :)

I killed the shortcut and disabled the feature, but it might be better if the mapping could be set to something else when not on a US keyboard.

**Comment by Thieum** [comment]
Same happens on a canadian french keyboard. It's been a long standing issue with PPT.

**Comment by olegtk** [comment]
Yeah, we know we should never use Ctrl+Alt based shortcuts. You can remap it via Tools/Options/Keyboard, but we also need to come up with a new shortcut and fix it.

**Comment by olegtk** [comment]
Oh, Align Assignments is not open sourced :(

**Comment by jbaute** [comment]
On a Belgian keyboard I'm thinking Ctrl-Shift-< would be a good shortcut.
It's not used AFAIK and you only need your left hand for it.

**Comment by justcla** [comment]
jbaute, can you please clarify this:

> What happens is that the shortcut for Align Assignments is set to Ctrl-Alt-^.
On a Belgian keyboard however, this is what you have to press to get an opening square bracket.

First, I'm confused about Align Assignments being set to Ctrl+Alt+^. Is that how it shows in Tools->Options->Keyboard? It's meant to be assigned to Ctrl+Alt+]
On the Belgian keyboard the ^ char is on the open square bracket key - [
And the close square bracket is on the $/* key immediately to the right.

Also, I was of the understanding that you had to press AltGr (the right Alt key) to activate those keys. Are you saying that you can get the open square bracket by pressing Ctrl+Alt+^  ?

**Comment by jbaute** [comment]
Yes, pressing Ctrl-Alt-^ is the same as AltGr-^, which gives you the "[" character.
These combo's both translate to modifier keys Ctrl & Alt, so they appear as exactly the same key-presses to VS.
So that keybinding blocks use of the "[" character on a Belgian keyboard. :)

**Comment by Thieum** [comment]
Here are the layouts:
- Belgian:
![Belgian Keyboard](https://upload.wikimedia.org/wikipedia/commons/9/93/Belgian_keyboard_layout.png)
- French Canadian:
![French Canadian Keyboard](http://www3.uakron.edu/modlang/french/images/kbd4.gif)

You can see the pattern with the ] requiring the alt-car to be pressed to be used.

**Comment by Thieum** [comment]
Successions of touches are nice to avoid special characters, like the Ctrl+K then C for comments that is built-in.

**Comment by justcla** [comment]
jbaute - but this is meant to be attached to the CLOSE bracket key, not the OPEN bracket key. So I can't figure out why your system has bound the command to a different key.

Are you sure it's bound to OPEN bracket?
What does it look like in the Tools->Options;Keyboard dialog?

**Comment by jbaute** [comment]
![vs2010shortcut](https://user-images.githubusercontent.com/10774709/27442782-74fb6190-5771-11e7-8630-47dbd39d7fa2.png)

I don't have the original binding anymore, but this is what it looked like.

**Comment by justcla** [comment]
Fascinating! Thanks for sharing.

We advice using the AltGr key for special characters, as we bind commands to Ctrl+Alt.
But this is only working for Ctrl+Alt+[A-Z,0-9]. We are looking to expand this to include Shift, and all keys - not just alpha numeric.
Hopefully that will solve the problem.

**Comment by justcla** [comment]
Notes on this topic:

Interesting how the shortcut gets mapped to a totally different key in the Belgian keyboard.
I note that the VSCT declares the following binding:

`<KeyBinding guid="guidAlignAssignmentsCmdSet" id="cmdidAlignAssignments" key1="VK_OEM_6" mod1="Control Alt" editor="guidVSStd97" />`

So it's using the VK_OEM_6 key, which explains why it jumps to a totally different key.

[Windows VK-Codes](https://msdn.microsoft.com/en-us/library/windows/desktop/dd375731(v=vs.85).aspx) shows it as:

> VK_OEM_6  (0xDD)
> Used for miscellaneous characters; it can vary by keyboard.
> For the US standard keyboard, the ']}' key
> 

But, clearly, it's mapped to the ^ [ key on Belgian keyboards.

**Comment by jbaute** [comment]
Interestign bug-hunt!  :)

**Comment by Thieum** [comment]
Still mapped to the ] key on french canadian keyboards in vs2017 15.8.9 / align assignments v15.0.5
